specreduce/extract.py

```diff
@@ -143,3 +145,158 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
+                    variance = image.uncertainty.array
+                elif image.uncertainty.uncertainty_type == 'std':
+                    # CCDData default uncertainties given as pure arrays to std
+                    # and logs a warning saying so upon object creation. should
+                    # we remind users again here?
+                    variance = image.uncertainty.array**2
+                else:
+                    # NOTE: last type is InverseVariance. should we take it?
+                    variance = image.uncertainty.array
+            else:
+                # ignore variance arg to focus on updating CCDData object
+                raise ValueError('image CCDData object lacks uncertainty')
+
+        else:
+            if any(arg is None for arg in (variance, mask, unit)):
+                raise ValueError('if image is a numpy array, the variance, '
+                                 'mask, and unit arguments must be specified. '
+                                 'consider wrapping that information into one '
+                                 'object by instead passing a CCDData image.')
+
+            if image.shape != variance.shape:
+                raise ValueError('image and variance shapes must match')
+            if image.shape != mask.shape:
+                raise ValueError('image and mask shapes must match')
+            if isinstance(unit, str):
+                unit = u.Unit(unit)
+
+            img = np.ma.array(image, mask=mask)
+
+        # co-add signal in each image column
+        ncols = img.shape[crossdisp_axis]
+        xd_pixels = np.arange(ncols)  # y plot dir / x spec dir
+        coadd = img.sum(axis=disp_axis) / ncols
+
+        # fit source profile, using Gaussian model as a template
+        # NOTE: could add argument for users to provide their own model
+        gauss_prof = models.Gaussian1D(amplitude=coadd.max(),
+                                       mean=coadd.argmax(), stddev=2)
+
+        # Fit extraction kernel to column with combined gaussian/bkgrd model
+        ext_prof = gauss_prof + bkgrd_prof
+        fitter = fitting.LevMarLSQFitter()
+        fit_ext_kernel = fitter(ext_prof, xd_pixels, coadd)
+
+        # use compound model to fit a kernel to each image column
+        # NOTE: infers Gaussian1D source profile; needs generalization for others
+        kernel_vals = []
+        norms = []
+        for col_pix in range(img.shape[disp_axis]):
+            # set gaussian model's mean as column's corresponding trace value
+            fit_ext_kernel.mean_0 = trace_object.trace[col_pix]
+            # NOTE: support for variable FWHMs forthcoming and would be here
+
+            # fit compound model to column
+            fitted_col = fit_ext_kernel(xd_pixels)
+
+            # save result and normalization
+            kernel_vals.append(fitted_col)
+            norms.append(fit_ext_kernel.amplitude_0
+                         * fit_ext_kernel.stddev_0 * np.sqrt(2*np.pi))
```

Comment:
Since we know the length in advance, it might be cheaper to create an empty numpy array and fill them, rather than append in the for loop.

Reply:
in my experience, appending to lists is the ~same performance-wise as filling a pre-allocated numpy array. appending to numpy arrays is the huge no-no. that's orders of magnitude slower than appending to lists.

Reply:
In case you missed it, I checked this earlier and found that "continuing to append to a list was a small percentage faster than filling arrays."

Reply:
ah, the threads are all mangled here. i do recall that now and that is consistent with what i've seen in other contexts and benchmarks.